tools/cldr-code/src/main/java/org/unicode/cldr/util/GlossonymConstructor.java

```diff
@@ -0,0 +1,124 @@
+package org.unicode.cldr.util;
+
+import com.ibm.icu.util.Output;
+import org.unicode.cldr.test.CheckNames;
+
+/**
+ * Automatically construct language names (glossonyms)
+ *
+ * Example: in German (de), for the path
+ *
+ *       //ldml/localeDisplayNames/languages/language[@type="ro_MD"]
+ *
+ * the value "Rumänisch (Republik Moldau)" is automatically constructed based on the code "ro_MD".
+ *
+ * The constructed value is a default if no preferable value is submitted or inherited.
+ * A different (non-constructed) value, such as "Moldauisch", may become the winning
+ * value instead of the constructed value.
+ */
+public class GlossonymConstructor {
+
+    /**
+     * Some paths with this prefix can get automatically constructed values
+     */
+    public static final String PATH_PREFIX = "//ldml/localeDisplayNames/languages/language[@type=\"";
+
+    /**
+     * The code such as "ro_MD" must contain an underscore, otherwise there is no constructed value.
+     * Underscore also serves as a clue for recognizing a value that is from code-fallback; for example,
+     * when the value "ro_MD" is inherited, the underscore implies it's a raw code ("bogus value") and should
+     * be replaced by a constructed value like "Rumänisch (Republik Moldau)"; but when the value "Moldauisch"
+     * (without underscore) is inherited, it should not be replaced by a constructed value
+     */
+    private static final String CODE_SEPARATOR = "_";
+
+    /**
+     * For "pathWhereFound" when the value is constructed.
+     * It is non-null to satisfy TestPathHeadersAndValues.
+     * It should not be treated as an actual path; for example, the
+     * Survey Tool Info Panel should not show a broken "Jump to original" link.
+     */
+    public static final String PSEUDO_PATH = "constructed";
+
+    /**
+     * Is the given path eligible for getting a constructed value?
+     *
+     * @param xpath the given path
+     * @return true if eligible
+     */
+    public static boolean pathIsEligible(String xpath) {
+        return xpath.startsWith(PATH_PREFIX) && xpath.contains(CODE_SEPARATOR);
+    }
+
+    /**
+     * Is the given value bogus, and therefore eligible for getting replaced by a constructed value?
+     *
+     * @param value the given value
+     * @return true if bogus
+     */
+    public static boolean valueIsBogus(String value) {
```

Comment:
We should be telling whether a value is bogus base on where it is found, not what it looks like. That is, if it comes from root or code fallback, then it's 'bogus'.

It looks like this is being called on the result from cldrFile.getName(type, true, altPicker); instead of guessing, we should enhance that method to return the information we need.

Reply:
I completely agree that should the goal. This PR avoids changing cldrFile.getName more than necessary. I think that realistically, any improvements to getName should be in subsequent PRs. This PR is already extensive. It fixes bugs, clears up discrepancies, and encapsulates key functionalities in a new class separate from CLDRFile. Adding changes to getName to this same PR would have too much risk of breaking code that depends on getName, unrelated to constructed language names (in the sense of this ticket). getName is extremely complex, crufty, and tightly coupled. It should be moved into its own new class separate from CLDRFile. That needs to be done very carefully as a task in its own right, not mixed in with other changes.

Tracking where a value is found also cries out for a new class. The way we currently pass around Output<String> pathWhereFound, Output<String> localeWhereFound -- is too cumbersome and ad hoc. We don't do it consistently, and it's not always clear what these strings should be. We do have some classes such as XMLSource.AliasLocation and XMLSource.SourceLocation, that might be the beginnings of something more generally useful. Such a class should encapsulate a notion of "where found" that's general enough to include constructed language names. As it stands, XMLSource doesn't know about constructed language names.

About valueIsBogus specifically, as it stands the test is only applied to values for a very limited set of paths, and the assumptions it makes about values with underscores and digits aren't new with this PR. I think it's relatively safe to merge this PR and follow up with enhancements such as "where found" tracking capability for getName.